src/server.rs

```diff
@@ -101,9 +99,93 @@ pub enum Event {
     },
 }
 
+/// Data structure containing data to be sent by server in messages containing rectangles.
+#[derive(Debug)]
+pub enum Rectangle<'a> {
+    /// Raw pixel data.
+    Raw { rect: Rect, pixel_data: &'a [u8] },
+
+    /// Rectangle in pixel data owned by client to copy data from.
+    CopyRect { src: Rect, dst: Rect },
+
+    /// Compressed pixel data.
+    Zrle { rect: Rect, zlib_data: &'a [u8] },
+
+    /// Data for drawing cursor.
+    SetCursor { size: (u16, u16), hotspot: (u16, u16), pixels: &'a [u8], mask_bits: &'a [u8] },
+
+    /// New size of framebuffer.
+    DesktopSize { width: u16, height: u16 },
+
+    /// Confirmation of support of pseudo-encoding.
+    Encoding { encoding: protocol::Encoding },
+}
+
+impl<'a> Rectangle<'a> {
+    fn check(&self, bytes_per_pixel: u16) -> Result<()> {
+        match *self {
+            Rectangle::Raw { ref rect, pixel_data } => {
+                let expected_num_bytes = rect.width as usize *
+                                         rect.height as usize *
+                                         bytes_per_pixel as usize;
+                if expected_num_bytes == pixel_data.len() {
+                    Ok(())
+                } else {
+                    let msg = format!("Expected data length for rectangle {:?} \
```

Comment:
This should just be a panic, no need to have a separate error type for a trivially avoidable contract violation.